File: backend/workline/knowledge/cache/memory.py
"""Thread-safe L1 in-memory bounded LRU cache with TTL support."""

from collections import OrderedDict
import threading
import time
from typing import Any, Dict, List, Optional, Tuple
from backend.workline.knowledge.cache.models import CacheMetadata


class MemoryCache:
    """Bounded in-process memory cache using OrderedDict for LRU eviction."""

    def __init__(self, max_entries: int = 1500):
        self.max_entries = max_entries
        self._lock = threading.RLock()
        self._cache: OrderedDict[str, Tuple[CacheMetadata, Any]] = OrderedDict()

    def get(self, key: str) -> Optional[Tuple[CacheMetadata, Any]]:
        with self._lock:
            if key not in self._cache:
                return None

            meta, data = self._cache[key]
            # Check expiration
            if meta.expires_at > 0 and time.time() > meta.expires_at:
                del self._cache[key]
                return None

            # Move to end for LRU refresh
            self._cache.move_to_end(key)
            return meta, data

    def set(self, key: str, meta: CacheMetadata, data: Any) -> None:
        with self._lock:
            if key in self._cache:
                del self._cache[key]
            elif len(self._cache) >= self.max_entries:
                # Evict least recently used (first item)
                self._cache.popitem(last=False)

            self._cache[key] = (meta, data)

    def has(self, key: str) -> bool:
        return self.get(key) is not None

    def delete(self, key: str) -> bool:
        with self._lock:
            if key in self._cache:
                del self._cache[key]
                return True
            return False

    def clear(self) -> None:
        with self._lock:
            self._cache.clear()

    def size(self) -> int:
        with self._lock:
            return len(self._cache)

    def clear_expired(self) -> int:
        with self._lock:
            now = time.time()
            expired_keys = [k for k, (meta, _) in self._cache.items() if meta.expires_at > 0 and now > meta.expires_at]
            for k in expired_keys:
                del self._cache[k]
            return len(expired_keys)

    def get_all_entries(self) -> List[Tuple[CacheMetadata, Any]]:
        with self._lock:
            return list(self._cache.values())

File: packages/knowledge/cache/memory.ts
/**
 * L1 Bounded In-Memory Cache with LRU eviction and TTL support.
 */

import { CacheEntry } from "./metadata";

export class MemoryCache {
  private map: Map<string, CacheEntry>;
  private readonly maxEntries: number;

  constructor(maxEntries: number = 1000) {
    this.maxEntries = maxEntries;
    this.map = new Map<string, CacheEntry>();
  }

  public get<T>(key: string): CacheEntry<T> | null {
    const entry = this.map.get(key);
    if (!entry) return null;

    // Check expiration
    if (entry.metadata.expiresAt > 0 && Date.now() > entry.metadata.expiresAt) {
      this.map.delete(key);
      return null;
    }

    // Refresh LRU order (delete & re-insert)
    this.map.delete(key);
    this.map.set(key, entry);

    return entry as CacheEntry<T>;
  }

  public set<T>(entry: CacheEntry<T>): void {
    const key = entry.metadata.cacheKey;
    if (this.map.has(key)) {
      this.map.delete(key);
    } else if (this.map.size >= this.maxEntries) {
      // Evict oldest entry
      const oldestKey = this.map.keys().next().value;
      if (oldestKey) {
        this.map.delete(oldestKey);
      }
    }
    this.map.set(key, entry);
  }

  public has(key: string): boolean {
    return this.get(key) !== null;
  }

  public delete(key: string): boolean {
    return this.map.delete(key);
  }

  public clear(): void {
    this.map.clear();
  }

  public size(): number {
    return this.map.size;
  }

  public entries(): CacheEntry[] {
    return Array.from(this.map.values());
  }

  public clearExpired(): number {
    const now = Date.now();
    let count = 0;
    for (const [key, entry] of this.map.entries()) {
      if (entry.metadata.expiresAt > 0 && now > entry.metadata.expiresAt) {
        this.map.delete(key);
        count++;
      }
    }
    return count;
  }
}
